Python/immutability.c

```diff
@@ -0,0 +1,376 @@
+                        goto nomemory;
+                    }
+                }
+            }
+        }
+
+        size = PySequence_Fast_GET_SIZE(f_code->co_consts);
+        for(Py_ssize_t i = 0; i < size; i++){
+            PyObject* value = PySequence_Fast_GET_ITEM(f_code->co_consts, i); // value.rc = x
+            if(!_Py_IsImmutable(value)){
+                if(PyCode_Check(value)){
+                    _Py_SetImmutable(value);
+
+                    if(PyList_Append(f_stack, value)){
+                        goto nomemory;
+                    }
+                }else{
+                    if(PyList_Append(frontier, value)){
+                        goto nomemory;
+                    }
+                }
+            }
+
+            if(check_globals && PyUnicode_Check(value)){
+                PyObject* name = value;
+                if(PyDict_Contains(globals, name)){
+                    value = PyDict_GetItem(globals, name);
+                    if(PyDict_SetItem(frozen_globals, name, value)){
+                        Py_DECREF(frozen_builtins);
+                        Py_DECREF(frozen_globals);
+                        Py_DECREF(f_stack);
+                        return NULL;
+                    }
+                }
+            }
+        }
+    }
+
+    Py_DECREF(f_stack);
+
+    if(check_globals){
+        size = 0;
+        if(f->func_closure != NULL)
+            size = PySequence_Fast_GET_SIZE(f->func_closure);
+
+        for(Py_ssize_t i=0; i < size; ++i){
+            PyObject* cellvar = PySequence_Fast_GET_ITEM(f->func_closure, i); // cellvar.rc = x
+            PyObject* value = PyCell_GET(cellvar); // value.rc = x
+
+            if(PyUnicode_Check(value)){
+                PyObject* name = value;
+                if(PyDict_Contains(globals, name)){
+                    value = PyDict_GetItem(globals, name);
+                    if(PyDict_SetItem(frozen_globals, name, value)){
+                        Py_DECREF(frozen_builtins);
+                        Py_DECREF(frozen_globals);
+                        return NULL;
+                    }
+                }
+            }
+        }
+    }
+
+    if(PyList_Append(frontier, frozen_globals)){
+        goto nomemory;
+    }
+
+    f->func_globals = frozen_globals;
+    Py_DECREF(globals);
+
+    if(PyList_Append(frontier, frozen_builtins)){
+        goto nomemory;
+    }
+
+    f->func_builtins = frozen_builtins;
+    Py_DECREF(builtins);
+
+    Py_RETURN_NONE;
+
+nomemory:
+    Py_XDECREF(frozen_builtins);
+    Py_XDECREF(frozen_globals);
+    Py_XDECREF(f_stack);
+    return PyErr_NoMemory();
+}
+
+static int freeze_visit(PyObject* obj, void* frontier)
+{
+    if(!_Py_IsImmutable(obj)){
+        if(PyList_Append((PyObject*)frontier, obj)){
+            PyErr_NoMemory();
+            return -1;
+        }
+    }
+
+    return 0;
+}
+
+PyObject* _Py_Freeze(PyObject* obj)
+{
+    if(_Py_IsImmutable(obj) && _Py_IsImmutable(Py_TYPE(obj))){
```

Comment:
Question: 
Do you have an example how it can happen that the object is immutable, but the type is mutable?

Reply:
There was an example previously, but it is defunct now. I don't think this is possible anymore. Let me think on it.